PhysicsTools/PatAlgos/plugins/PackedCandidateMuonSelectorProducer.cc

```diff
@@ -0,0 +1,119 @@
+namespace pat {
+
+  class PackedCandidateMuonSelectorProducer : public edm::stream::EDProducer<> {
+    typedef edm::ValueMap<bool> BoolMap;
+
+  public:
+    explicit PackedCandidateMuonSelectorProducer(const edm::ParameterSet& iConfig)
+        : muonToken_(consumes<pat::MuonCollection>(iConfig.getParameter<edm::InputTag>("muons"))),
+          candidateToken_(
+              consumes<edm::View<pat::PackedCandidate> >(iConfig.getParameter<edm::InputTag>("candidates"))),
+          muonSelectors_(iConfig.getParameter<std::vector<std::string> >("muonSelectors")),
+          muonIDs_(iConfig.getParameter<std::vector<std::string> >("muonIDs")) {
+      for (const auto& sel : muonSelectors_) {
+        produces<BoolMap>(sel);
+      }
+      for (const auto& sel : muonIDs_) {
+        muonIDMap_[sel].reset(new StringCutObjectSelector<pat::Muon>("passed('" + sel + "')"));
+        produces<BoolMap>(sel);
+      }
+    }
+    ~PackedCandidateMuonSelectorProducer() override{};
+
+    void produce(edm::Event&, const edm::EventSetup&) override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    edm::EDGetTokenT<pat::MuonCollection> muonToken_;
+    edm::EDGetTokenT<edm::View<pat::PackedCandidate> > candidateToken_;
+    std::vector<std::string> muonSelectors_;
+    std::vector<std::string> muonIDs_;
+    std::map<std::string, std::unique_ptr<StringCutObjectSelector<pat::Muon> > > muonIDMap_;
+  };
+
+}  // namespace pat
+
+void pat::PackedCandidateMuonSelectorProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<pat::MuonCollection> muons;
+  edm::Handle<edm::View<pat::PackedCandidate> > candidates;
+
+  iEvent.getByToken(muonToken_, muons);
+  iEvent.getByToken(candidateToken_, candidates);
+
+  const auto& nCand = candidates->size();
+
+  std::map<std::string, std::vector<bool> > muonSelMap;
+  for (const auto& sel : muonSelectors_) {
+    muonSelMap[sel] = std::vector<bool>(nCand, false);
+  }
+  for (const auto& sel : muonIDs_) {
+    muonSelMap[sel] = std::vector<bool>(nCand, false);
+  }
+
+  // loop over muons
+  for (const auto& muon : *muons) {
+    // ignore muons without high purity inner track
+    if (muon.innerTrack().isNull() || !muon.innerTrack()->quality(reco::TrackBase::qualityByName("highPurity")))
+      continue;
+
+    // find candidate associated to muon
+    const auto& muonTrack = *muon.innerTrack();
+    for (size_t i = 0; i < nCand; i++) {
+      const auto& cand = candidates->refAt(i);
+      // ignore neutral candidates or without track
+      if (cand->charge() == 0 || !cand->hasTrackDetails())
+        continue;
+
+      // check if candidate and muon are compatible
+      const auto& candTrack = cand->pseudoTrack();
+      if (muonTrack.charge() == candTrack.charge() && muonTrack.numberOfValidHits() == candTrack.numberOfValidHits() &&
+          std::abs(muonTrack.eta() - candTrack.eta()) < 1E-3 && std::abs(muonTrack.phi() - candTrack.phi()) < 1E-3 &&
```

Comment:
Is this phi subtraction safe in case of wraparound? I think one might also use a `deltaR` method from the framework: https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaR.h
In addition, perhaps it can be considered to put these cutoffs 1E-3, 1E-2 at the top of the file as a static const, or otherwise as a configurable parameter in a visible spot?

Reply:
Thanks for the suggestions. I replaced the deltaPhi and deltaEta for a deltaR, and changed the cutoffs to be configurable parameters.